rec.c

```diff
@@ -1,5 +1,61 @@
 #include "rec.h"
 
+bool
+brute_rec_gen_handler (task_t *task, void *context)
+{
+  (void)task; // to suppress "unused parameter" warning
+  rec_state_t *state = (rec_state_t *)context;
+  swapcontext (&state->contexts[1], &state->contexts[0]);
+  
+  return (false);
+}
+
+void
+brute_rec_gen_helper (config_t *config, rec_state_t *state)
+{
+  brute_rec_wrapper (state->task, config, brute_rec_gen_handler, state);
+  state->cancelled = true;
+}
+
+void
+rec_state_init (rec_state_t *state, task_t *task, config_t *config)
+{
+  state->task = task;
+  state->cancelled = false;
+
+  getcontext (&state->contexts[0]);
+  state->contexts[1] = state->contexts[0];
+  state->contexts[1].uc_stack.ss_sp = &state->stack;
+  state->contexts[1].uc_stack.ss_size = sizeof (state->stack);
+  state->contexts[1].uc_link = &state->contexts[0];
+  makecontext (&state->contexts[1], (void (*) (void))brute_rec_gen_helper, 2,
+               config, state);
+
+  swapcontext (&state->contexts[0], &state->contexts[1]);
+}
+
+bool
+rec_state_next (rec_state_t *state)
+{
+  swapcontext (&state->contexts[0], &state->contexts[1]);
```

Comment:
canceled also needs to be checked before context swap. No need to to iterate if iterator is finished.